### ISSUE TITLE
InvalidCastException when using SagaData has a CorrelationID with the name 'ID' 

### ISSUE THREAD

**Issue by HEskandari** [Particular/NServiceBus#5631]
When the correlation ID of the saga is called 'ID' a runtime cast exception is thrown which is incorrect.

Imagine the case of the following SagaData:

```c#
public class DownloadProcessData : ContainSagaData
{
    public long ID { get; set; }
}
``` 

While this code is valid and it does not conflict with the 'Id' property in the base class, at runtime user gets an error as follows:

```
System.InvalidCastException: Unable to cast object of type 'System.Int64' to type 'System.Guid'.
   at NServiceBus.PropertySagaFinder`1.Find(IBuilder builder, SagaFinderDefinition finderDefinition, SynchronizedStorageSession storageSession, ContextBag cont
ext, Object message)
   at NServiceBus.SagaPersistenceBehavior.Invoke(IInvokeHandlerContext context, Func`2 next)
   at NServiceBus.LoadHandlersConnector.Invoke(IIncomingLogicalMessageContext context, Func`2 stage)
   at NServiceBus.ScheduledTaskHandlingBehavior.Invoke(IIncomingLogicalMessageContext context, Func`2 next)
   at NServiceBus.InvokeSagaNotFoundBehavior.Invoke(IIncomingLogicalMessageContext context, Func`2 next)
   at NServiceBus.DeserializeMessageConnector.Invoke(IIncomingPhysicalMessageContext context, Func`2 stage)
   at NServiceBus.InvokeAuditPipelineBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next)
   at NServiceBus.ProcessingStatisticsBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next)
   at NServiceBus.TransportReceiveToPhysicalMessageConnector.Invoke(ITransportReceiveContext context, Func`2 next)
   at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext)
   at NServiceBus.Transport.SqlServer.ReceiveStrategy.TryProcessingMessage(Message message, TransportTransaction transportTransaction)
   at NServiceBus.Transport.SqlServer.ProcessWithNativeTransaction.TryProcess(Message message, TransportTransaction transportTransaction)
Exception details:
        Message ID: bb63b5a0-7736-480d-8c9c-abac0
```

The root-case is that we have special-cased the ID property and there's no distinction between 'Id' and 'ID' when storing the data in the underlying persistence (in this case SQL Persistence). See the related code [here](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Sagas/PropertySagaFinder.cs#L37).

A proper way of handling this would have been doing type-checking on the value and throw a more meaningful exception if the ID does not match.

**Comment by ramonsmits** [comment]
@HEskandari Where did this issue originate? Having more background info might help to prioritize but currently, it seems more like a theoretical enhancement.

**Comment by DavidBoike** [comment]
Saga analyzers to be released in NServiceBus 7.7 will include an analyzer to prevent mapping a correlation id of Id or any alternate casing id, ID, or even iD. It will be a warning in NServiceBus 7.7 and an error in NServiceBus 8.0. This is the PR targeting master for NServiceBus 8: https://github.com/Particular/NServiceBus/pull/6234
